File: acmicpc/A01300.java
package acmicpc;

import java.io.*;

public class A01300 {
    public static int solution(int n, int required) {
        int low = 1, high = required;
        while (low <= high) {
            int mid = low + (high - low) / 2, count = 0;
            for (int i = 1; i <= n; i++) {
                count += Math.min(n, mid / i);
            }
            if (count < required)
                low = mid + 1;
            else
                high = mid - 1;
        }
        return low;
    }

    public void reader() throws NumberFormatException, IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        int n = Integer.parseInt(br.readLine());
        int k = Integer.parseInt(br.readLine());
        System.out.println(solution(n, k));
    }

    public static void main(String[] args) throws NumberFormatException, IOException {
        A01300 test = new A01300();
        test.reader();
    }
}

File: acmicpc/A01300.py
from sys import stdin as s


def solution(n: int, k: int) -> None:
    low, high = 1, k
    while low <= high:
        mid = (low + high) // 2
        count = sum(min(n, mid // i) for i in range(1, n + 1))
        if count < k:
            low = mid + 1
        else:
            high = mid - 1
    print(low)


def reader():
    n = int(s.readline().strip())
    k = int(s.readline().strip())
    solution(n, k)


if __name__ == '__main__':
    reader()
